kf/bpf.go

```diff
@@ -1233,3 +1178,266 @@ func ValidatePath(filePath string, destination string) (string, error) {
+			}
+		} else if tmpMetricsMap.KeySize() == 8 {
+			var k int64
+			switch tmpMetricsMap.ValueSize() {
+			case 1:
+				var v int8
+				err = tmpMetricsMap.Update(unsafe.Pointer(&k), unsafe.Pointer(&v), 0)
+			case 2:
+				var v int16
+				err = tmpMetricsMap.Update(unsafe.Pointer(&k), unsafe.Pointer(&v), 0)
+			case 4:
+				var v int32
+				err = tmpMetricsMap.Update(unsafe.Pointer(&k), unsafe.Pointer(&v), 0)
+			case 8:
+				var v int64
+				err = tmpMetricsMap.Update(unsafe.Pointer(&k), unsafe.Pointer(&v), 0)
+			default:
+				log.Error().Msgf("unsupported key type int64 and value size - %d", tmpMetricsMap.ValueSize())
+			}
+		}
+		if err != nil {
+			return fmt.Errorf("update hash map element failed for map name %s error %v", tmpMap.Name, err)
+		}
+	}
+	return nil
+}
+
+// Isloaded - Method verifies bpf program is loaded or not
+// Here it checks prog ID is valid and active
+func (b *BPF) Isloaded() bool {
+	ebpfProg, err := ebpf.NewProgramFromID(b.ProgID)
+	if err != nil && errors.Is(err, os.ErrNotExist) {
+		log.Debug().Msgf("Isloaded - %s is not loaded or invalid program id %d", b.Program.Name, uint32(b.ProgID))
+		return false
+	}
+	defer ebpfProg.Close()
+	return true
+}
+
+func (b *BPF) StartUserProgram(ifaceName, direction string, chain bool) error {
+	cmd := filepath.Join(b.FilePath, b.Program.CmdStart)
+	// Validate
+	if err := assertExecutable(cmd); err != nil {
+		return fmt.Errorf("no executable permissions on %s - error %v", b.Program.CmdStart, err)
+	}
+
+	args := make([]string, 0, len(b.Program.StartArgs)<<1)
+	args = append(args, "--iface="+ifaceName)     // attaching to interface
+	args = append(args, "--direction="+direction) // direction xdpingress or ingress or egress
+
+	if chain && b.ProgMapCollection == nil {
+		if len(b.PrevMapNamePath) > 1 {
+			args = append(args, "--map-name="+b.PrevMapNamePath)
+		}
+	}
+
+	if len(b.hostConfig.BPFLogDir) > 1 {
+		args = append(args, "--log-dir="+b.hostConfig.BPFLogDir)
+	}
+
+	if len(b.Program.RulesFile) > 1 && len(b.Program.Rules) > 1 {
+		fileName, err := b.createUpdateRulesFile(direction)
+		if err == nil {
+			args = append(args, "--rules-file="+fileName)
+		}
+	}
+
+	for k, val := range b.Program.StartArgs {
+		if v, ok := val.(string); !ok {
+			err := fmt.Errorf("start args is not a string for the ebpf program %s", b.Program.Name)
+			log.Error().Err(err).Msgf("failed to convert start args value into string for program %s", b.Program.Name)
+			return err
+		} else {
+			args = append(args, "--"+k+"="+v)
+		}
+	}
+
+	log.Info().Msgf("BPF Program start command : %s %v", cmd, args)
+	b.Cmd = execCommand(cmd, args...)
+	if err := b.Cmd.Start(); err != nil {
+		log.Info().Err(err).Msgf("user mode BPF program failed - %s", b.Program.Name)
+		return fmt.Errorf("failed to start : %s %v", cmd, args)
+	}
+	if !b.Program.UserProgramDaemon {
+		log.Info().Msgf("no user mode BPF program - %s No Pid", b.Program.Name)
```

Comment:
```suggestion
		log.Info().Msgf("no user program - %s No Pid", b.Program.Name)
```
A "BPF program" is different from the user program.